oak/server/oak_runtime.cc

```diff
@@ -30,116 +30,113 @@
-                << "}." << dest_port;
-      MessageChannel::ChannelHalves halves = MessageChannel::Create();
-      if (dest_node == log_node) {
-        // Remember the write half of logging channel for future re-use.
-        logging_channel = halves.write.get();
-      }
-
-      src_node->AddNamedChannel(src_port, absl::make_unique<ChannelHalf>(std::move(halves.write)));
-      dest_node->AddNamedChannel(dest_port, absl::make_unique<ChannelHalf>(std::move(halves.read)));
-    }
+  // Create a gRPC pseudo-Node.
+  const std::string grpc_name = kGrpcNodeName;
+  LOG(INFO) << "Create gRPC pseudo-Node named {" << grpc_name << "}";
+  std::unique_ptr<OakGrpcNode> grpc_node = OakGrpcNode::Create(grpc_name);
+  grpc_node_ = grpc_node.get();  // borrowed copy
+  nodes_[grpc_name] = std::move(grpc_node);
+
+  // Create the initial Application Node.
+  std::string node_name;
+  OakNode* app_node = CreateNode(config.initial_node(), &node_name);
+  if (app_node == nullptr) {
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Failed to create initial Oak Node");
   }
+  LOG(INFO) << "Created Wasm node named {" << node_name << "}";
+
+  // Create an initial channel from gRPC pseudo-Node to Application Node.
+  // Both of the initial nodes have exactly one registered handle.
+  MessageChannel::ChannelHalves halves = MessageChannel::Create();
+  Handle grpc_handle =
+      grpc_node_->AddChannel(absl::make_unique<ChannelHalf>(std::move(halves.write)));
+  Handle app_handle = app_node->AddChannel(absl::make_unique<ChannelHalf>(std::move(halves.read)));
+  LOG(INFO) << "Created initial channel from Wasm node {" << grpc_name << "}." << grpc_handle
+            << " to {" << node_name << "}." << app_handle;
 
   return grpc::Status::OK;
 }
 
-std::string OakRuntime::NextNodeName(const std::string& contents) {
-  absl::MutexLock lock(&mu_);
-  int index = next_index_[contents]++;
-  return absl::StrCat(contents, "-", index);
+std::string OakRuntime::NextNodeName(const std::string& config) {
+  int index = next_index_[config]++;
+  return absl::StrCat(config, "-", index);
 }
 
-bool OakRuntime::CreateWasmNode(const std::string& contents, std::unique_ptr<ChannelHalf> half,
-                                std::string* node_name) {
-  if (wasm_contents_.count(contents) != 1) {
-    LOG(ERROR) << "failed to find Wasm bytecode Node contents with name " << contents;
-    return false;
+// Create (but don't start) a new Node instance.  Return a borrowed pointer to
+// the new Node (or nullptr on failure).
+OakNode* OakRuntime::CreateNode(const std::string& config, std::string* node_name) {
+  std::string name = NextNodeName(config);
+  std::unique_ptr<OakNode> node;
+
+  if (wasm_config_.count(config) > 0) {
+    LOG(INFO) << "Create Wasm node named {" << name << "}";
+    const std::string* module_bytes = wasm_config_[config].get();
+    node = WasmNode::Create(this, name, *module_bytes);
```

Comment:
This looks slightly inconsistent with the logging and storage ones. Should we have a ::Create for all of them?

Reply:
It is a bit inconsistent, but that's because creation of a pseudo-Node can't fail (it's provided by the runtime) but creation of a Wasm Node can (the user might have passed in invalid Wasm code).  So the former can use a constructor, but the latter needs a factory method.

So the inconsistency reflects the untrustworthiness of users&hellip;